File: frontend/src/api/bookdrop.ts
import { api } from './client';
import type { BookDetail } from './books';

// Mirrors internal/handler/bookdrop.go bookdropDTO.
export type BookDropState =
  | 'discovered'
  | 'processing'
  | 'ready'
  | 'failed'
  | 'imported'
  | 'rejected';

export type BookDropItem = {
  id: string;
  filename: string;
  path: string;
  fileSize: number;
  format: string;
  state: BookDropState;
  progress: number;
  errorMsg?: string;
  title?: string;
  author?: string;
  description?: string;
  language?: string;
  hasCover: boolean;
  coverMime?: string;
  bookId?: string;
  discoveredAt: string;
  updatedAt: string;
};

export async function fetchBookDrop(): Promise<BookDropItem[]> {
  const { items } = await api<{ items: BookDropItem[] }>('/api/v1/bookdrop');
  return items;
}

export async function approveBookDrop(
  id: string,
  libraryId?: string,
): Promise<BookDetail> {
  const { book } = await api<{ book: BookDetail }>(
    `/api/v1/bookdrop/${id}/approve`,
    {
      method: 'POST',
      body: libraryId ? JSON.stringify({ libraryId }) : undefined,
    },
  );
  return book;
}

export async function rejectBookDrop(id: string): Promise<void> {
  await api<void>(`/api/v1/bookdrop/${id}/reject`, { method: 'POST' });
}

export const bookdropQueryKey = ['bookdrop'] as const;

export const bookdropCoverUrl = (id: string) => `/api/v1/bookdrop/${id}/cover`;

File: internal/model/bookdrop.go
package model

import "time"

// BookDropState is the lifecycle of a single file in the ingest pipeline.
type BookDropState string

const (
	BookDropDiscovered BookDropState = "discovered" // queued, not yet processed
	BookDropProcessing BookDropState = "processing" // worker is running
	BookDropReady      BookDropState = "ready"      // metadata extracted, awaiting review
	BookDropFailed     BookDropState = "failed"     // processor returned an error
	BookDropImported   BookDropState = "imported"   // user approved → books row created
	BookDropRejected   BookDropState = "rejected"   // user dismissed
)

// BookDropItem mirrors a row in bookdrop_items.
type BookDropItem struct {
	ID           string
	Path         string
	FileSize     int64
	Format       string
	State        BookDropState
	Progress     int
	ErrorMsg     string
	Title        string
	Author       string
	Description  string
	Language     string
	HasCover     bool
	CoverMime    string
	BookID       *string
	DiscoveredAt time.Time
	UpdatedAt    time.Time
	// ContentHash is the SHA-256 digest of the file content, computed
	// during ingest. nil (or empty) until Task 9 fills it.
	ContentHash []byte
}

// IsTerminal reports whether no further transitions are expected.
func (s BookDropState) IsTerminal() bool {
	return s == BookDropImported || s == BookDropRejected
}

// NeedsReview reports whether the UI should surface Approve/Reject actions.
func (s BookDropState) NeedsReview() bool {
	return s == BookDropReady || s == BookDropFailed
}
